src/operator.rs

```diff
@@ -157,10 +157,12 @@ impl<'a, 'mir, 'tcx> EvalContextExt<'tcx> for super::MiriEvalContext<'a, 'mir, '
                 assert_eq!(size as u64, self.pointer_size().bytes());
                 let bits = bits as u64;
 
-                // Case I: Comparing with NULL
-                if bits == 0 {
-                    // Test if the ptr is in-bounds. Then it cannot be NULL.
-                    // Even dangling pointers cannot be NULL.
+                // Case I: Comparing with small integers
+                if bits <= 1 {
+                    // Test if the ptr is in-bounds. Then it cannot be a small integer.
+                    // Even dangling pointers cannot be small integers.
+                    // The only exception is wasm and some embedded devices, but that's a different
+                    // story: https://github.com/rust-lang/rust/issues/57897
```

Comment:
I don't like this.  Enabling such platform-specific assumptions should be somehow, at the very least, tied to the platform -- as in, depending on the target we should do this or not.  I think we should also have a flag ("strict mode" or so) to not make such assumptions.

For platforms where we make this assumption, probably it applies at least to `bits < 4096`, doesn't it?